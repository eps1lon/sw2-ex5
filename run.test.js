const child_process = require("child_process");
const fs = require("fs").promises;
const lodash = require("lodash");
const path = require("path");
const rimraf = require("rimraf");
const { promisify } = require("util");
const xmlParser = require("fast-xml-parser");

const exec = promisify(child_process.exec);

const useModified = false;
/**
 * modified outputs as _impl.wsdl
 * function removes this suffix in order to diff the files
 */
function getModifiedFileName(filename) {
  const basename = path.basename(filename, ".wsdl");
  const expectedBasename = `${basename}_impl`;

  return path.join(path.dirname(filename), `${expectedBasename}.wsdl`);
}

const tests = [
  [
    "widgetPrice",
    {
      options: [
        `-o tmp/wp.wsdl`,
        `-l"http://localhost:8080/axis/services/WidgetPrice"`,
        `-n "urn:Example6"`,
        `-p"samples.userguide.example"`,
        `"urn:Example6"`,
        `org.apache.axis.wsdl.Java2WSDL`
      ],
      files: ["tmp/wp.wsdl"]
    }
  ],
  [
    "widgetPriceOtherFile",
    {
      options: [
        `-o tmp/wp-other.wsdl`,
        `-l"http://localhost:8080/axis/services/WidgetPrice"`,
        `-n "urn:Example6"`,
        `-p"samples.userguide.example"`,
        `"urn:Example6"`,
        `org.apache.axis.wsdl.Java2WSDL`
      ],
      files: ["tmp/wp-other.wsdl"]
    }
  ],
  ["usage", { options: [], files: [] }],
  ["help", { options: ["--help"], files: [] }],
  [
    "widgetPriceStyleDOCUMENT",
    {
      options: [
        `-o tmp/wp-other.wsdl`,
        `-l"http://localhost:8080/axis/services/WidgetPrice"`,
        `-n "urn:Example6"`,
        `-p"samples.userguide.example"`,
        `--style DOCUMENT`,
        `"urn:Example6"`,
        `org.apache.axis.wsdl.Java2WSDL`
      ],
      files: []
    }
  ],
  [
    "widgetPriceStyleWSDL",
    {
      options: [
        `-o tmp/wp-other.wsdl`,
        `-l"http://localhost:8080/axis/services/WidgetPrice"`,
        `-n "urn:Example6"`,
        `-p"samples.userguide.example"`,
        `--style RPC`,
        `"urn:Example6"`,
        `org.apache.axis.wsdl.Java2WSDL`
      ],
      files: []
    }
  ],
  [
    "ex6-widgetPrice",
    {
      options: [
        `-o tmp/ex6-widgetPrice.wsdl`,
        `-p"samples.userguide.example6"`,
        `-l"http://localhost:8080/axis/services/WidgetPrice"`,
        `-n "urn:Example6"`,
        `samples.userguide.example6.WidgetPrice`
      ],
      files: ["tmp/ex6-widgetPrice.wsdl"]
    }
  ],
  [
    "ex5-BeanService",
    {
      options: [
        `-o tmp/ex5-beanservice.wsdl`,
        `-p"samples.userguide.example5"`,
        `-l"http://localhost:8080/axis/services/ex5"`,
        `-n "urn:Example5"`,
        `samples.userguide.example5.BeanService`
      ],
      files: ["tmp/ex5-beanservice.wsdl"]
    }
  ]
];

beforeEach(async () => {
  rimraf.sync("tmp");
  await fs.mkdir("tmp");
});

it.each(tests)("%s matches", async (id, { options, files }) => {
  const binary = `java -cp "axis-1_4/lib/*:axis-1_4/build/classes" org.apache.axis.wsdl.Java2WSDL`;

  const input = options.join(" ");
  const command = `${binary} ${input}`;

  let stdout;
  let stderr;
  let error;
  try {
    ({ stdout, stderr } = await exec(command));
  } catch (caughtError) {
    error = caughtError;
  }

  expect(error).toMatchSnapshot("actual error");
  expect(stdout).toMatchSnapshot("stdout");
  expect(stderr).toMatchSnapshot("stderr");
  for (const expectedFile of files) {
    const actualFileName = useModified
      ? getModifiedFileName(expectedFile)
      : expectedFile;
    const xml = await fs.readFile(actualFileName, { encoding: "utf8" });
    const wsdl = xmlParser.parse(xml, { ignoreAttributes: false });
    const stableWsdl = deepSortStable(wsdl);

    expect(stableWsdl).toMatchSnapshot(`file ${expectedFile}`);
  }
});

/**
 * sorts objects by their keys
 * sorts arrays by their JSON.stringified value
 * all deep
 */
function deepSortStable(obj) {
  if (Array.isArray(obj)) {
    return obj
      .map(deepSortStable)
      .sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));
  }
  if (typeof obj === "object" && obj !== null) {
    return Object.fromEntries(
      Object.entries(obj)
        .map(([key, value]) => {
          return [key, deepSortStable(value)];
        })
        .sort(([a], [b]) => {
          return a.localeCompare(b);
        })
    );
  }
  return obj;
}
