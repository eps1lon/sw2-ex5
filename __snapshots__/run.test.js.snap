// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ex6-widgetPrice matches 1`] = `
"- Unable to find required classes (javax.activation.DataHandler and javax.mail.internet.MimeMultipart). Attachment support is disabled.
"
`;

exports[`ex6-widgetPrice matches 2`] = `""`;

exports[`ex6-widgetPrice matches 3`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<wsdl:definitions targetNamespace=\\"urn:Example6\\" xmlns:apachesoap=\\"http://xml.apache.org/xml-soap\\" xmlns:impl=\\"urn:Example6\\" xmlns:intf=\\"urn:Example6\\" xmlns:soapenc=\\"http://schemas.xmlsoap.org/soap/encoding/\\" xmlns:wsdl=\\"http://schemas.xmlsoap.org/wsdl/\\" xmlns:wsdlsoap=\\"http://schemas.xmlsoap.org/wsdl/soap/\\" xmlns:xsd=\\"http://www.w3.org/2001/XMLSchema\\">


   <wsdl:message name=\\"getWidgetPriceRequest\\">

      <wsdl:part name=\\"in0\\" type=\\"soapenc:string\\"/>

   </wsdl:message>

   <wsdl:message name=\\"setWidgetPriceResponse\\">

   </wsdl:message>

   <wsdl:message name=\\"setWidgetPriceRequest\\">

      <wsdl:part name=\\"in0\\" type=\\"soapenc:string\\"/>

      <wsdl:part name=\\"in1\\" type=\\"soapenc:string\\"/>

   </wsdl:message>

   <wsdl:message name=\\"getWidgetPriceResponse\\">

      <wsdl:part name=\\"getWidgetPriceReturn\\" type=\\"soapenc:string\\"/>

   </wsdl:message>

   <wsdl:portType name=\\"WidgetPrice\\">

      <wsdl:operation name=\\"setWidgetPrice\\" parameterOrder=\\"in0 in1\\">

         <wsdl:input message=\\"impl:setWidgetPriceRequest\\" name=\\"setWidgetPriceRequest\\"/>

         <wsdl:output message=\\"impl:setWidgetPriceResponse\\" name=\\"setWidgetPriceResponse\\"/>

      </wsdl:operation>

      <wsdl:operation name=\\"getWidgetPrice\\" parameterOrder=\\"in0\\">

         <wsdl:input message=\\"impl:getWidgetPriceRequest\\" name=\\"getWidgetPriceRequest\\"/>

         <wsdl:output message=\\"impl:getWidgetPriceResponse\\" name=\\"getWidgetPriceResponse\\"/>

      </wsdl:operation>

   </wsdl:portType>

   <wsdl:binding name=\\"WidgetPriceSoapBinding\\" type=\\"impl:WidgetPrice\\">

      <wsdlsoap:binding style=\\"rpc\\" transport=\\"http://schemas.xmlsoap.org/soap/http\\"/>

      <wsdl:operation name=\\"setWidgetPrice\\">

         <wsdlsoap:operation soapAction=\\"\\"/>

         <wsdl:input name=\\"setWidgetPriceRequest\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:input>

         <wsdl:output name=\\"setWidgetPriceResponse\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:output>

      </wsdl:operation>

      <wsdl:operation name=\\"getWidgetPrice\\">

         <wsdlsoap:operation soapAction=\\"\\"/>

         <wsdl:input name=\\"getWidgetPriceRequest\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:input>

         <wsdl:output name=\\"getWidgetPriceResponse\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:output>

      </wsdl:operation>

   </wsdl:binding>

   <wsdl:service name=\\"WidgetPriceService\\">

      <wsdl:port binding=\\"impl:WidgetPriceSoapBinding\\" name=\\"WidgetPrice\\">

         <wsdlsoap:address location=\\"http://localhost:8080/axis/services/WidgetPrice\\"/>

      </wsdl:port>

   </wsdl:service>

</wsdl:definitions>
"
`;

exports[`help matches 1`] = `undefined`;

exports[`help matches 2`] = `undefined`;

exports[`usage matches 1`] = `undefined`;

exports[`usage matches 2`] = `undefined`;

exports[`widgetPrice matches 1`] = `
"- Unable to find required classes (javax.activation.DataHandler and javax.mail.internet.MimeMultipart). Attachment support is disabled.
"
`;

exports[`widgetPrice matches 2`] = `""`;

exports[`widgetPrice matches 3`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<wsdl:definitions targetNamespace=\\"urn:Example6\\" xmlns:apachesoap=\\"http://xml.apache.org/xml-soap\\" xmlns:impl=\\"urn:Example6\\" xmlns:intf=\\"urn:Example6\\" xmlns:soapenc=\\"http://schemas.xmlsoap.org/soap/encoding/\\" xmlns:wsdl=\\"http://schemas.xmlsoap.org/wsdl/\\" xmlns:wsdlsoap=\\"http://schemas.xmlsoap.org/wsdl/soap/\\" xmlns:xsd=\\"http://www.w3.org/2001/XMLSchema\\">

 <wsdl:types>
  <schema targetNamespace=\\"urn:Example6\\" xmlns=\\"http://www.w3.org/2001/XMLSchema\\">
   <import namespace=\\"http://schemas.xmlsoap.org/soap/encoding/\\"/>
   <complexType name=\\"ArrayOf_soapenc_string\\">
    <complexContent>
     <restriction base=\\"soapenc:Array\\">
      <attribute ref=\\"soapenc:arrayType\\" wsdl:arrayType=\\"soapenc:string[]\\"/>
     </restriction>
    </complexContent>
   </complexType>
  </schema>
 </wsdl:types>

   <wsdl:message name=\\"mainRequest\\">

      <wsdl:part name=\\"args\\" type=\\"impl:ArrayOf_soapenc_string\\"/>

   </wsdl:message>

   <wsdl:message name=\\"mainResponse\\">

   </wsdl:message>

   <wsdl:portType name=\\"Java2WSDL\\">

      <wsdl:operation name=\\"main\\" parameterOrder=\\"args\\">

         <wsdl:input message=\\"impl:mainRequest\\" name=\\"mainRequest\\"/>

         <wsdl:output message=\\"impl:mainResponse\\" name=\\"mainResponse\\"/>

      </wsdl:operation>

   </wsdl:portType>

   <wsdl:binding name=\\"WidgetPriceSoapBinding\\" type=\\"impl:Java2WSDL\\">

      <wsdlsoap:binding style=\\"rpc\\" transport=\\"http://schemas.xmlsoap.org/soap/http\\"/>

      <wsdl:operation name=\\"main\\">

         <wsdlsoap:operation soapAction=\\"\\"/>

         <wsdl:input name=\\"mainRequest\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:input>

         <wsdl:output name=\\"mainResponse\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:output>

      </wsdl:operation>

   </wsdl:binding>

   <wsdl:service name=\\"Java2WSDLService\\">

      <wsdl:port binding=\\"impl:WidgetPriceSoapBinding\\" name=\\"WidgetPrice\\">

         <wsdlsoap:address location=\\"http://localhost:8080/axis/services/WidgetPrice\\"/>

      </wsdl:port>

   </wsdl:service>

</wsdl:definitions>
"
`;

exports[`widgetPriceOtherFile matches 1`] = `
"- Unable to find required classes (javax.activation.DataHandler and javax.mail.internet.MimeMultipart). Attachment support is disabled.
"
`;

exports[`widgetPriceOtherFile matches 2`] = `""`;

exports[`widgetPriceOtherFile matches 3`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<wsdl:definitions targetNamespace=\\"urn:Example6\\" xmlns:apachesoap=\\"http://xml.apache.org/xml-soap\\" xmlns:impl=\\"urn:Example6\\" xmlns:intf=\\"urn:Example6\\" xmlns:soapenc=\\"http://schemas.xmlsoap.org/soap/encoding/\\" xmlns:wsdl=\\"http://schemas.xmlsoap.org/wsdl/\\" xmlns:wsdlsoap=\\"http://schemas.xmlsoap.org/wsdl/soap/\\" xmlns:xsd=\\"http://www.w3.org/2001/XMLSchema\\">

 <wsdl:types>
  <schema targetNamespace=\\"urn:Example6\\" xmlns=\\"http://www.w3.org/2001/XMLSchema\\">
   <import namespace=\\"http://schemas.xmlsoap.org/soap/encoding/\\"/>
   <complexType name=\\"ArrayOf_soapenc_string\\">
    <complexContent>
     <restriction base=\\"soapenc:Array\\">
      <attribute ref=\\"soapenc:arrayType\\" wsdl:arrayType=\\"soapenc:string[]\\"/>
     </restriction>
    </complexContent>
   </complexType>
  </schema>
 </wsdl:types>

   <wsdl:message name=\\"mainRequest\\">

      <wsdl:part name=\\"args\\" type=\\"impl:ArrayOf_soapenc_string\\"/>

   </wsdl:message>

   <wsdl:message name=\\"mainResponse\\">

   </wsdl:message>

   <wsdl:portType name=\\"Java2WSDL\\">

      <wsdl:operation name=\\"main\\" parameterOrder=\\"args\\">

         <wsdl:input message=\\"impl:mainRequest\\" name=\\"mainRequest\\"/>

         <wsdl:output message=\\"impl:mainResponse\\" name=\\"mainResponse\\"/>

      </wsdl:operation>

   </wsdl:portType>

   <wsdl:binding name=\\"WidgetPriceSoapBinding\\" type=\\"impl:Java2WSDL\\">

      <wsdlsoap:binding style=\\"rpc\\" transport=\\"http://schemas.xmlsoap.org/soap/http\\"/>

      <wsdl:operation name=\\"main\\">

         <wsdlsoap:operation soapAction=\\"\\"/>

         <wsdl:input name=\\"mainRequest\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:input>

         <wsdl:output name=\\"mainResponse\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:output>

      </wsdl:operation>

   </wsdl:binding>

   <wsdl:service name=\\"Java2WSDLService\\">

      <wsdl:port binding=\\"impl:WidgetPriceSoapBinding\\" name=\\"WidgetPrice\\">

         <wsdlsoap:address location=\\"http://localhost:8080/axis/services/WidgetPrice\\"/>

      </wsdl:port>

   </wsdl:service>

</wsdl:definitions>
"
`;

exports[`widgetPriceStyleDOCUMENT matches 1`] = `undefined`;

exports[`widgetPriceStyleDOCUMENT matches 2`] = `undefined`;

exports[`widgetPriceStyleDOCUMENT matches 3`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<wsdl:definitions targetNamespace=\\"urn:Example6\\" xmlns:apachesoap=\\"http://xml.apache.org/xml-soap\\" xmlns:impl=\\"urn:Example6\\" xmlns:intf=\\"urn:Example6\\" xmlns:soapenc=\\"http://schemas.xmlsoap.org/soap/encoding/\\" xmlns:wsdl=\\"http://schemas.xmlsoap.org/wsdl/\\" xmlns:wsdlsoap=\\"http://schemas.xmlsoap.org/wsdl/soap/\\" xmlns:xsd=\\"http://www.w3.org/2001/XMLSchema\\">

 <wsdl:types>
  <schema targetNamespace=\\"urn:Example6\\" xmlns=\\"http://www.w3.org/2001/XMLSchema\\">
   <import namespace=\\"http://schemas.xmlsoap.org/soap/encoding/\\"/>
   <complexType name=\\"ArrayOf_soapenc_string\\">
    <complexContent>
     <restriction base=\\"soapenc:Array\\">
      <attribute ref=\\"soapenc:arrayType\\" wsdl:arrayType=\\"soapenc:string[]\\"/>
     </restriction>
    </complexContent>
   </complexType>
  </schema>
 </wsdl:types>

   <wsdl:message name=\\"mainRequest\\">

      <wsdl:part name=\\"args\\" type=\\"impl:ArrayOf_soapenc_string\\"/>

   </wsdl:message>

   <wsdl:message name=\\"mainResponse\\">

   </wsdl:message>

   <wsdl:portType name=\\"Java2WSDL\\">

      <wsdl:operation name=\\"main\\" parameterOrder=\\"args\\">

         <wsdl:input message=\\"impl:mainRequest\\" name=\\"mainRequest\\"/>

         <wsdl:output message=\\"impl:mainResponse\\" name=\\"mainResponse\\"/>

      </wsdl:operation>

   </wsdl:portType>

   <wsdl:binding name=\\"WidgetPriceSoapBinding\\" type=\\"impl:Java2WSDL\\">

      <wsdlsoap:binding style=\\"rpc\\" transport=\\"http://schemas.xmlsoap.org/soap/http\\"/>

      <wsdl:operation name=\\"main\\">

         <wsdlsoap:operation soapAction=\\"\\"/>

         <wsdl:input name=\\"mainRequest\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:input>

         <wsdl:output name=\\"mainResponse\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:output>

      </wsdl:operation>

   </wsdl:binding>

   <wsdl:service name=\\"Java2WSDLService\\">

      <wsdl:port binding=\\"impl:WidgetPriceSoapBinding\\" name=\\"WidgetPrice\\">

         <wsdlsoap:address location=\\"http://localhost:8080/axis/services/WidgetPrice\\"/>

      </wsdl:port>

   </wsdl:service>

</wsdl:definitions>
"
`;

exports[`widgetPriceStyleWSDL matches 1`] = `undefined`;

exports[`widgetPriceStyleWSDL matches 2`] = `undefined`;

exports[`widgetPriceStyleWSDL matches 3`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<wsdl:definitions targetNamespace=\\"urn:Example6\\" xmlns:apachesoap=\\"http://xml.apache.org/xml-soap\\" xmlns:impl=\\"urn:Example6\\" xmlns:intf=\\"urn:Example6\\" xmlns:soapenc=\\"http://schemas.xmlsoap.org/soap/encoding/\\" xmlns:wsdl=\\"http://schemas.xmlsoap.org/wsdl/\\" xmlns:wsdlsoap=\\"http://schemas.xmlsoap.org/wsdl/soap/\\" xmlns:xsd=\\"http://www.w3.org/2001/XMLSchema\\">

 <wsdl:types>
  <schema targetNamespace=\\"urn:Example6\\" xmlns=\\"http://www.w3.org/2001/XMLSchema\\">
   <import namespace=\\"http://schemas.xmlsoap.org/soap/encoding/\\"/>
   <complexType name=\\"ArrayOf_soapenc_string\\">
    <complexContent>
     <restriction base=\\"soapenc:Array\\">
      <attribute ref=\\"soapenc:arrayType\\" wsdl:arrayType=\\"soapenc:string[]\\"/>
     </restriction>
    </complexContent>
   </complexType>
  </schema>
 </wsdl:types>

   <wsdl:message name=\\"mainRequest\\">

      <wsdl:part name=\\"args\\" type=\\"impl:ArrayOf_soapenc_string\\"/>

   </wsdl:message>

   <wsdl:message name=\\"mainResponse\\">

   </wsdl:message>

   <wsdl:portType name=\\"Java2WSDL\\">

      <wsdl:operation name=\\"main\\" parameterOrder=\\"args\\">

         <wsdl:input message=\\"impl:mainRequest\\" name=\\"mainRequest\\"/>

         <wsdl:output message=\\"impl:mainResponse\\" name=\\"mainResponse\\"/>

      </wsdl:operation>

   </wsdl:portType>

   <wsdl:binding name=\\"WidgetPriceSoapBinding\\" type=\\"impl:Java2WSDL\\">

      <wsdlsoap:binding style=\\"rpc\\" transport=\\"http://schemas.xmlsoap.org/soap/http\\"/>

      <wsdl:operation name=\\"main\\">

         <wsdlsoap:operation soapAction=\\"\\"/>

         <wsdl:input name=\\"mainRequest\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:input>

         <wsdl:output name=\\"mainResponse\\">

            <wsdlsoap:body encodingStyle=\\"http://schemas.xmlsoap.org/soap/encoding/\\" namespace=\\"urn:Example6\\" use=\\"encoded\\"/>

         </wsdl:output>

      </wsdl:operation>

   </wsdl:binding>

   <wsdl:service name=\\"Java2WSDLService\\">

      <wsdl:port binding=\\"impl:WidgetPriceSoapBinding\\" name=\\"WidgetPrice\\">

         <wsdlsoap:address location=\\"http://localhost:8080/axis/services/WidgetPrice\\"/>

      </wsdl:port>

   </wsdl:service>

</wsdl:definitions>
"
`;
